src/epanet.c

```diff
@@ -3166,6 +3302,352 @@ int DLLEXPORT ENgetaveragepatternvalue(int index, EN_API_FLOAT_TYPE *value)
   return(0);
 }
 
+int DLLEXPORT ENsetlinktype(char *id, EN_LinkType toType) {
+  int i;
+  EN_LinkType fromType;
+  
+  if ( !Openflag ) return(102);
+  
+/* Check if a link with the id exists */
+  if ( ENgetnodeindex(id, &i) != 0 ) return(215);
+  
+/* Get the current type of the link */
+  ENgetnodetype(i, fromType);
+  if(fromType == toType) return(0);
+  
+  
+/* Change link from Pipe */
+  if(toType <= EN_PIPE) {
+    Npipes++;
+  }
+  else if(toType == EN_PUMP) {
+    Npumps++;
+    Pump[Npumps].Link = i;
+  } else {
+    Nvalves++;
+    Valve[Nvalves].Link = i;
+  }
+  
+  if(fromType <= EN_PIPE) {
+    Npipes--;
+  } else if(fromType == EN_PUMP) {
+    Npumps--;
+    
+  }
+  
+}
+
+int  DLLEXPORT  ENaddnode(char *id, EN_NodeType nodeType)

```

Comment:
see note for links also. Edge case is specifying an ID of length greater than `MAXID` (31) characters, which would truncate the ID silently. Should probably return an error if the user attempts this.